io/common/include/detray/io/geometry_writer.hpp

```diff
@@ -0,0 +1,193 @@
+/** Detray library, part of the ACTS project (R&D line)
+ *
+ * (c) 2023 CERN for the benefit of the ACTS project
+ *
+ * Mozilla Public License Version 2.0
+ */
+
+#pragma once
+
+// Project include(s)
+#include "detray/io/payloads.hpp"
+
+// System include(s)
+#include <string>
+#include <vector>
+
+namespace detray {
+
+/// @brief Abstract base class for tracking geometry writers
+template <class detector_t>
+class geometry_writer {
+
+    public:
+    /// All writers must define a file name
+    geometry_writer() = delete;
+
+    /// File gets created with a fixed @param extension
+    geometry_writer(const std::string& ext) : m_file_extension{ext} {}
+
+    /// Default destructor
+    virtual ~geometry_writer() {}
+
+    /// Writes the tracking geometry of the detector to a file with a given name
+    virtual void write(const detector_t&, const std::string&) = 0;
```

Comment:
Similar to my previous comment, this function could receive an abstract [std::ostream](https://en.cppreference.com/w/cpp/io/basic_ostream) for the output file.

And I also don't really like that this class is not receiving the same type of object that `detray::geometry_reader` is returning. :frowning: Would it not be possible to make them operate on the same type? :thinking:

Reply:
Roughly speaking, this one turns a detector into and intermediate ```std::vector<volume_payload>```, while the ```geometry_reader``` should do the reverse. But since the detector builder infrastructure is not complete, yet, I can only return the intermediate representation at this point from the reader, not a detector.